### ISSUE TITLE
Multi Site: How to map a custom domain to a Camaleon-app-subdomain on Heroku 

### ISSUE THREAD

**Issue by mayinx** [owen2345/camaleon-cms#543]
Hi there!

I've setup a multi-site solution with Camaleon CMS on Heroku (1 main-/default-site like `www.mydomain.de`+ n further sites accessible via subdomains/keys like `fu.mydomain.de`, `bar.mydomain.de` etc...  )

For this to work I defined (via DNSimple)  `mydomain.de` as an ALIAS for `myapp.herokuapp.com` and enabled wildcard subdomains for `mydomain.de` by defining a CNAME record which sets `*.mydomain.de`  as alias for `myapp.herokuapp.com`  and by adding  ` *.mydomain.de ` as domain name to my Heroku-app (via `heroku domains:add *.mydomain.de`). To round it up I defined another CNAME record to set `www.mydomain.de` as another alias for `myapp.herokuapp.com` & added `www.mydomain.de` as well to my Heroku app (`heroku domains:add mydomain.de` ). 

Here's the DNS-setup (DNS Records on DNSimple for `mydomain.de`):

Type | Name | Content
------------ | ------------- | ------------- 
`ALIAS` | `mydomain.de` | `myapp.herokuapp.com`	 
`CNAME` | `*.mydomain.de` | `myapp.herokuapp.com`		  
`CNAME` | `www.mydomain.de` | `myapp.herokuapp.com`
`TXT` | `mydomain.de` | `ALIAS for myapp.herokuapp.com`
`NS`	| `mydomain.de` | `ns4.dnsimple.com`		
`NS`	| `mydomain.de` | `ns2.dnsimple.com`		
`NS`	| `mydomain.de` | `ns1.dnsimple.com`		
`NS`	| `mydomain.de` | `ns3.dnsimple.com`

And here are the custom domains I added on Heroku for `myapp.herokuapp.com`:

Domain Name | DNS Target
------------ | -------------
`*.mydomain.de` | `myapp.herokuapp.com`
`mydomain.de` | `myapp.herokuapp.com`
`www.mydomain.de` | `myapp.herokuapp.com`

That works great so far....

## TL;DR? ;-)

Now Id'like to map external domains to those subdomains - to allow clients to use their own domains instead of the subdomains of the Camaleon-app - i.e. `www.fusdomain.de ` shall be mapped to  `fu.mydomain.de` and `www.barsdomain.de` shall be mapped to `bar.mydomain.de` and so forth ...

What is the best way to achieve this? (for testing purposes I tried to map one of my domains to a subdomain (`CNAME  www.testdomain.de  -> somesubdomain.mydomain.de`) and added it to heroku (`heroku domains:add www.testdomain.de`) - but that didn't work, `www.test.de` simply resolved to the main/default-site  `www.mydomain.de` instead of the desired subdomain...)

Cheers
Mayinx

**Comment by owen2345** [comment]
Please check this:
http://camaleon.tuzitio.com/documentation/category/40756-uncategorized/how.html

**Comment by FlowerWrong** [comment]
@owen2345 it is a 500 page.

**Comment by owen2345** [comment]
The url is working well, tested from many computers and also using web proxies.
![screen shot 2016-11-25 at 05 58 40](https://cloud.githubusercontent.com/assets/1148415/20621209/d068cbd2-b2d3-11e6-8511-762370f37972.png)

**Comment by mazharoddin** [comment]
@owen2345, yesterday it was not working for me aswell, but today I am able to access.

**Comment by owen2345** [comment]
@mazharoddin yes it will works too!

**Comment by mazharoddin** [comment]
ok, thanks.

**Comment by mazharoddin** [comment]
hi @owen2345,
I tried above helper file, but for some reason it didn't worked for me. 
I update application as below, i tried these on production server.

```
def shoplid_current_site_helper(args)
    if !args[:site].present? && request.present?
      args[:site] = case request.original_url.to_s.parse_domain
      when 'othersubromain.mywebsite.com'
        CamaleonCms::Site.find_by_slug('localhost:3000').decorate
      when 'mywesbite.com'
        CamaleonCms::Site.find_by_slug('localhost:4000').decorate
    end
   end
  end

```
also tried as below.

def shoplid_current_site_helper(args)
    if !args[:site].present? && request.present?
      args[:site] = case request.original_url.to_s.parse_domain
      when 'mywebsite.com, othersubdomain.mywebsite.com'
        CamaleonCms::Site.find_by_slug('localhost:3000').decorate
      when 'othersubdomain.com'
        CamaleonCms::Site.find_by_slug('localhost:4000').decorate
    end
   end
  ```
but either way it didn't worked for me. I restarted ngnix, unicorn etc.

**Comment by owen2345** [comment]
@mazharoddin 
the helper name should be the same, try like this:
```
def cama_current_site_helper(args)
  args[:site] = case request.original_url.to_s.parse_domain
      when 'mywebsite.com, othersubdomain.mywebsite.com'
        CamaleonCms::Site.find_by_slug('localhost:3000').decorate
      when 'othersubdomain.com'
        CamaleonCms::Site.find_by_slug('localhost:4000').decorate
      else
        args[:site]
    end
end
```

**Comment by mazharoddin** [comment]
thanks, but still its not working for me ..may be I am missing something. I will try and let you know, thanks.

**Comment by mayinx** [comment]
Thanks a lot, folks - finally I got the time to try this out, once I'm through I report back to you guys ...  

@mazharoddin: I didn't figure it all out yet - but I had the same problem than you when it came to the helper method: When I override `cama_current_site_helper` in `app/helpers/application_helper.rb` it's never called - but if I override the method in `app/apps/themes/default/default_helper.rb` instead it's invoked - so if I'm not mistaken `cama_current_site_helper` doesn't seem to belong (anymore?) in `application_helper.rb` as suggested in the documentation (is that correct @owen2345 ?)   ...

**Comment by owen2345** [comment]
You are right @mayinx, I am sorry.
For some reason is not calling/loading the application helper.
Anyway I updated the document with the right solution:
http://camaleon.tuzitio.com/documentation/category/40756-uncategorized/how.html
Regards,

**Comment by mazharoddin** [comment]
@owen2345, I am still not able to get it work. 
@mayinx  is its working for you now?

**Comment by owen2345** [comment]
@mazharoddin could you please print something to test if the function is acceded?

**Comment by mazharoddin** [comment]
Sure @owen2345 ..I will do it. thanks

**Comment by mazharoddin** [comment]
Looks like no where its getting called.

**Comment by owen2345** [comment]
it is strange:
![custom-map-sites](https://cloud.githubusercontent.com/assets/1148415/20754159/0b097a92-b6e0-11e6-91ea-5748bb2adae5.png)
Don't forget to restart server

**Comment by mazharoddin** [comment]
no, its not working, may be something wrong from my end.

**Comment by mayinx** [comment]
@mazharoddin  - yeah - for me it's working great (thanks @owen2345 !)  - now I can map external / custom domains to the virtual subdomains of my Camaleon-CMS main-/default-site - and I hadn't even to change the original slug of the virtual subdomain in question ... I can now access the site via the regular subdomain-notation (`somesubdomain.mydomain.de`) AND via the custom domain (`testdomain.de`) - here's my `camaleon_mapsites.rb`, in case it helps: 


```ruby
# config/initializers/camaleon_mapsites.rb
#
# To inspect what's going on use stuff like the following inside the helper method:
# puts "--- Cama::Site.all.pluck(:slug): '#{ Cama::Site.all.pluck(:slug)}'"#.red
# puts "--- args[:site]: '#{args[:site]}'"#.red
# puts "--- request.original_url: '#{request.original_url}'"#.red
# puts "--- request.original_url.to_s.parse_domain: '#{request.original_url.to_s.parse_domain}'"#.red
Rails.application.config.to_prepare do
  CamaleonCms::SiteHelper.module_eval do
    def cama_current_site_helper(args)    
      if !args[:site].present? && request.present?
        args[:site] =
          case request.original_url.to_s.parse_domain
            when 'testdomain.de'
              CamaleonCms::Site.find_by_slug('somesubdomain').decorate             
          end
      end
    end
  end
end
``` 

Don't now if it's related but perhaps it helps you guys to fix this: I'm on ruby v 2.1.2, rails v 4.2.1 and camalleon-cms v 2.3.6 - and I added the following at the end of my gemfile (found that in an external Camaleon-CMS-tutorial somewhere if I'm not mistaken...):  

```ruby
#################### Camaleon CMS include all gems for plugins and themes #################### 
require './lib/plugin_routes' 
instance_eval(PluginRoutes.draw_gems)
```

Hope that helps?

By the way: I'm writing a post about how I set everything up to get a multi site solution started incl. all DNS Settings etc. - once it's finished I gonna post the link here ...

**Comment by mazharoddin** [comment]
@mayinx  Its nice to know that its working for you, thanks for your sample code and details ..really appreciate it...I will re-try again and let you know.

**Comment by nasim31** [comment]
Any hope to run this code by fixing  "mapping custom domains"  helper file issue.

**Comment by owen2345** [comment]
What about this?
http://camaleon.tuzitio.com/documentation/category/139779-faqs/how.html
